llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp

```diff
@@ -2868,3 +2889,888 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+        ValueBits::ValueBit::CreateXORBit(CRCInputRefBit, DataInputRefBit);
+  } else {
+    RefCheckBit = CRCInputRefBit;
+  }
+
+  if (!RefCheckBit->equals(CheckBit)) {
+    LLVM_DEBUG(dbgs() << DEBUG_TYPE
+                      << " CRCRegonize: Cannot verify check bit!\n"
+                      << *RefCheckBit << "\n"
+                      << *CheckBit << "\n");
+    return false;
+  }
+
+  ValueBits *CRCOutBitsIfOne = CRCOutBitsPred->getIfTrue();
+  ValueBits *CRCOutBitsIfZero = CRCOutBitsPred->getIfFalse();
+  if (checkZero)
+    std::swap(CRCOutBitsIfZero, CRCOutBitsIfOne);
+
+  // Now construct ValueBits that would be the result of crc for one iteration.
+  // That is, a shift and then xor if [M/L]SB is 1.
+  ValueBits *CRCValueBits = nullptr;
+  Result = ValueMap.find(CRC.CRCInput);
+  if (Result == ValueMap.end()) {
+    CRCValueBits = new ValueBits(CRC.CRCInput, CRCSize);
+  } else {
+    CRCValueBits = Result->second;
+  }
+  uint64_t GeneratorPolynomial =
+      CRC.BitReversed ? reverseBits(CRC.Polynomial, CRCSize) : CRC.Polynomial;
+  ValueBits Polynomial(GeneratorPolynomial, CRCSize);
+
+  // Case where the MSB/LSB of the data is 0
+  ValueBits *IfZero = CRC.BitReversed ? ValueBits::LShr(CRCValueBits, 1)
+                                      : ValueBits::Shl(CRCValueBits, 1);
+
+  // Case where the MSB/LSB of the data is 1
+  ValueBits *IfOne = ValueBits::Xor(IfZero, &Polynomial);
+
+  if (!IfZero->equals(CRCOutBitsIfZero)) {
+    LLVM_DEBUG(dbgs() << DEBUG_TYPE << " CRCRegonize: Not Equal!\n"
+                      << *IfZero << *CRCOutBitsPred->getIfFalse());
+    return false;
+  }
+  if (!IfOne->equals(CRCOutBitsIfOne)) {
+    LLVM_DEBUG(dbgs() << DEBUG_TYPE << " CRCRegonize: Not Equal!\n"
+                      << *IfOne << *CRCOutBitsPred->getIfTrue());
+    return false;
+  }
+
+  LLVM_DEBUG(dbgs() << DEBUG_TYPE << " CRCRegonize: This looks like crc!\n");
+
+  writeTableBasedCRCOneByte(CRC);
+
+  return true;
+}
+
+std::optional<LoopIdiomRecognize::CRCInfo>
+LoopIdiomRecognize::looksLikeCRC(const SCEV *BECount) {
+  // Initial checks to see if this loop looks like CRC:
+  // - Inner most loop
+  // - One block
+  // - One exit
+  // - Iteration count is 8
+
+  // Check if this is inner most loop
+  if (!CurLoop->isInnermost())
+    return std::nullopt;
+
+  // Since we are far enough in the optimization pipeline that small branches
+  // will have been folded into Select instructions, if we have branches we are
+  // unlikely to be CRC. To reduce complexity, only consider single-block loops
+  // for CRC recognition
+  if (CurLoop->getBlocks().size() > 1) {
+    LLVM_DEBUG(dbgs() << DEBUG_TYPE " CRCRegonize: Loops with more than one"
+                      << "block are unsupported\n");
+    return std::nullopt;
+  }
+
+  // Ensure one exit block
+  const BasicBlock *ExitBlock = CurLoop->getExitBlock();
+  if (!ExitBlock)
+    return std::nullopt;
+
+  // Check iteration count is 8
+  const SCEV *TripCountSCEV =
```

Comment:
It's less code to use ScalarEvolution::getSmallConstantTripCount, but we already have the backedge count so I don't know what's better.